argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,417 @@
+
+            cls_to_ninst[class_name] = gt_filtered.shape[0]
+        return cls_to_accum, cls_to_ninst
+
+    def assign(self, dts: np.ndarray, gts: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """Attempt assignment of each detection to a ground truth label.
+
+        Args:
+            dts: Detections of shape (N,).
+            gts: Ground truth labels of shape (M,).
+
+        Returns:
+            metrics: Matrix of true/false positive concatenated with true positive errors (N, K + S) where K is the number
+                of true positive thresholds used for AP computation and S is the number of true positive errors.
+            scores: Corresponding scores for the true positives/false positives (N,)
+        """
+
+        # Ensure the number of boxes considered per class is at most `MAX_NUM_BOXES`.
+        if dts.shape[0] > MAX_NUM_BOXES:
+            dts = dts[:MAX_NUM_BOXES]
+
+        n_threshs = len(self.detection_cfg.affinity_threshs)
+        metrics = np.zeros((dts.shape[0], n_threshs + N_TP_ERRORS))
+
+        # Set the true positive metrics to np.nan since error is undefined on false positives.
+        metrics[:, n_threshs : n_threshs + N_TP_ERRORS] = np.nan
+        if gts.shape[0] == 0:
+            return metrics
+
+        affinity_matrix = compute_affinity_matrix(dts, gts, self.detection_cfg.affinity_fn_type)
+
+        # Get the GT label for each max-affinity GT label, detection pair.
+        gt_matches = affinity_matrix.argmax(axis=1)[np.newaxis, :]
+
+        # The affinity matrix is an N by M matrix of the detections and ground truth labels respectively.
+        # We want to take the corresponding affinity for each of the initial assignments using `gt_matches`.
+        # The following line grabs the max affinity for each detection to a ground truth label.
+        affinities = np.take_along_axis(affinity_matrix.T, gt_matches, axis=0).squeeze(0)
+
+        # Find the indices of the "first" detection assigned to each GT.
+        unique_gt_matches, unique_dt_matches = np.unique(gt_matches, return_index=True)
+        for i, thresh in enumerate(self.detection_cfg.affinity_threshs):
+
+            # `tp_mask` may need to be defined differently with other affinities
+            tp_mask = affinities[unique_dt_matches] > -thresh
+            metrics[unique_dt_matches, i] = tp_mask
+
+            # Only compute true positive error when `thresh` is equal to the tp threshold.
+            is_tp_thresh = thresh == self.detection_cfg.tp_thresh
+            # Ensure that there are true positives of the respective class in the frame.
+            has_true_positives = np.count_nonzero(tp_mask) > 0
+
+            if is_tp_thresh and has_true_positives:
+                dt_tp_indices = unique_dt_matches[tp_mask]
+                gt_tp_indices = unique_gt_matches[tp_mask]
+
+                dt_df = pd.DataFrame([dt.__dict__ for dt in dts[dt_tp_indices]])
```

Comment:
could add an inline comment just stating that from N object label records, we form N x D table, where D attributes in ObjectLabelRecord